core/modules/wpublish/QueriesAndChunks.cc

```diff
@@ -0,0 +1,533 @@
+        LOGS(_log, LOG_LVL_DEBUG, QueryIdHelper::makeIdStr(qId) << " was not found by removeQueryFrom");
+        return removedList;
+    }
+    lock.unlock();
+
+    // Remove Tasks from their scheduler put them on 'removedList', but only if their Scheduler is the same
+    // as 'sched' or if sched == nullptr.
+    auto& taskMap = query->second->_taskMap;
+    std::vector<wbase::Task::Ptr> taskList;
+    {
+        std::lock_guard<std::mutex> taskLock(query->second->_qStatsMtx);
+        for (auto const& elem : taskMap) {
+            taskList.push_back(elem.second);
+        }
+    }
+
+    for(auto const& task : taskList) {
+        auto taskSched = task->getTaskScheduler();
+        if (taskSched != nullptr && (taskSched == sched || sched == nullptr)) {
+            // removeTask will only return the task if it was found on the
+            // queue for its scheduler and removed.
+            auto removedTask = taskSched->removeTask(task);
+            if (removedTask != nullptr) {
+                removedList.push_back(removedTask);
+            }
+        }
+    }
+    return removedList;
+}
+
+
+/// Add the duration to the statistics for the table. Create a statistics object if needed.
+/// @return the statistics for the table.
+ChunkTableStats::Ptr ChunkStatistics::add(std::string const& scanTableName, double minutes) {
+    std::pair<std::string, ChunkTableStats::Ptr> ele(scanTableName, nullptr);
+    std::unique_lock<std::mutex> ul(_tStatsMtx);
+    auto res = _tableStats.insert(ele);
+    auto iter = res.first;
+    if (res.second) {
+        iter->second = std::make_shared<ChunkTableStats>(_chunkId, scanTableName);
+    }
+    ul.unlock();
+    iter->second->addTaskFinished(minutes);
+    return iter->second;
+}
+
+
+/// @return the statistics for a table. nullptr if the table is not found.
+ChunkTableStats::Ptr ChunkStatistics::getStats(std::string const& scanTableName) const {
+    std::lock_guard<std::mutex> g(_tStatsMtx);
+    auto iter = _tableStats.find(scanTableName);
+    if (iter != _tableStats.end()) {
+        return iter->second;
+    }
+    return nullptr;
+}
+
+
+/// Use the duration of the last Task completed to adjust the average completion time.
+void ChunkTableStats::addTaskFinished(double minutes) {
+    std::lock_guard<std::mutex> g(_cStatsMtx);
+    ++_tasksCompleted;
+    if (_tasksCompleted > 1) {
+        _avgCompletionTime = (_avgCompletionTime*_weightAvg + minutes*_weightNew)/_weightSum;
+    } else {
+        _avgCompletionTime = minutes;
+    }
+    LOGS(_log, LOG_LVL_DEBUG, "ChkId=" << _chunkId << ":tbl=" << _scanTableName
+         << " completed=" << _tasksCompleted
+         << " avgCompletionTime=" << _avgCompletionTime);
+}
+
+
+
+std::uint64_t ChunkTableStats::getTasksCompleted() {
+    std::lock_guard<std::mutex> g(_cStatsMtx);
+    return _tasksCompleted;
+}
+
+
+double ChunkTableStats::getAvgCompletionTime() {
+    std::lock_guard<std::mutex> g(_cStatsMtx);
+    return _avgCompletionTime;
+}
+
```

Comment:
methods `getTasksCompleted()` and `getAvgCompletionTime()` each lock `_cStatsMtx` but return only piece of the data protected by that mutex. Could their independent use result in inconsistent data being returned to caller?